ouroboros-consensus/docs/HardFork.md

```diff
@@ -0,0 +1,211 @@
+
+### Moment of the hard fork
+
+Similar to the Byron ledger, the Shelley ledger provides a "current protocol
+version", but it is a two-tuple (not a three-tuple), containing only a
+`hard fork` component and `soft fork` component:
+
+```haskell
+_protocolVersion :: (Natural, Natural)
+```
+
+in `PParams` (currently, module `PParams` in
+`chain-and-ledger/executable-spec/src/PParams.hs`).
+
+The hard fork from Shelley to its successor (Goguen?) will be initiated
+once the hard fork component of this version gets incremented.
+
+### The update mechanism for the protocol version
+
+The update mechanism in Shelley is simpler than it is in Byron. There is no
+distinction between votes and proposals: to "vote" for a proposal one merely
+submits the exact same proposal. There is also no separate endorsement step.
+The procedure is as follows:
+
+1. Proposals (change parameter `x` to value `y` are submitted to the chain)
+2. "Voting" (submitting of proposals) ends `2k` slots before the end of the
+   epoch (that `2k` might need to be adjusted pending more detailed analysis
+   from the researchers -- the spec records this as a single specification
+   parameter (not protocol parameter!) called `slotsPrior`).
+3. By the end of the epoch, if a majority (for some definition of majority) is
+   reached for _all_ proposed parameters, then _all_ parameters take their new
+   value at the start of the next epoch. If no majority is reached on _any one_
+   of the parameters, then _none_ of the parameters change. The next epoch is
+   always started with a clean slate, proposals from the previous epoch that
+   didn't make it are discarded. This all-or-nothing approach means that nodes
+   can indicate "I'm willing to change `A` in this way, but only if `B` changes
+   in that way".
+
+The protocol version itself is also considered to be merely another parameter,
+and parameters can change _without_ changing the protocol version, although
+a convention _could_ be established that the protocol version must change if
+any of the parameters do; but the specification itself does not mandate this.
+
+### Initiating the hard fork
+
+The timing of the hard fork in Shelley is different to the one in Byron;
+in Byron, we _first_ vote and then wait for people to get ready, in Shelley
+it is the other way around.
+
+(Core?) node operators will want to know that a (significant?) majority of the
+core nodes is ready (supports the hard fork) before initiating it. To make this
+visible, Shelley blocks contain a protocol version. This is not related to the
+current protocol version as reported by the ledger state (`_protocolVersion` as
+discussed in the previous section), but it is the _maximum_ protocol version
+that the node which produced that block can support.
+
+Once we see blocks from all or nearly all core nodes with the `hard fork`
+component of their protocol version equal to the post-hard-fork value, nodes
+will submit their votes to initiate the hard fork.
```

Comment:
This is not entirely clear to me. What does submitting a vote means here? Submitting a proposal with the version change?

Reply:
Yup, indeed. Will reword.